api/accesslog.proto

```diff
@@ -0,0 +1,138 @@
+syntax = "proto3";
+
+package envoy.api.v2;
+
+import "api/base.proto";
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+message EnvoyAccessLog {
+  // The request method, ie, "GET", "HEAD", "POST"
+  string request_method = 1;
+
+  // The time we started servicing this request
+  google.protobuf.Timestamp start_time = 2;
+
+  // Incoming protocol variation spoken
+  enum Protocol {
+    PROTOCOL_UNSPECIFIED = 0;
+    HTTP10 = 1;
+    HTTP11 = 2;
+    HTTP2 = 3;
+  }
+  Protocol protocol_variant = 3;
+
+  // This enum defines the various things that may have occurred while
+  // processing a request.
+  enum ResponseFlag {
+    // Local server healthcheck failed.
+    FAILED_LOCAL_HEALTHCHECK = 0;
+    // No healthy upstream.
+    NO_HEALTHY_UPSTREAM = 1;
+    // Request timeout on upstream.
+    UPSTREAM_REQUEST_TIMEOUT = 2;
+    // Local codec level reset was sent on the stream.
+    LOCAL_RESET = 3;
+    // Remote codec level reset was received on the stream.
+    UPSTREAM_REMOTE_RESET = 4;
+    // Local reset by a connection pool due to an initial connection failure.
+    UPSTREAM_CONNECTION_FAILURE = 5;
+    // If the stream was locally reset due to connection termination.
+    UPSTREAM_CONNECTION_TERMINATION = 6;
+    // The stream was reset because of a resource overflow.
+    UPSTREAM_OVERFLOW = 7;
+    // No route found for a given request.
+    NO_ROUTE_FOUND = 8;
+    // Request was delayed before proxying.
+    DELAY_INJECTED = 9;
+    // Abort with error code was injected.
+    FAULT_INJECTED = 10;
+    // Request was ratelimited locally by rate limit filter.
+    RATE_LIMITED = 11;
+  }
+  // Status flags about the response.
+  repeated ResponseFlag response_flags = 4;
+
+  // The remote (upstream) host, as an IPv4 or IPv6 string.
+  // IPv6 addresses should be stored in canonical (compressed) format.
+  string upstream_host = 5;
+
+  // The Upstream Cluster that the upstream host belongs to.
+  string upstream_cluster = 6;
+
+  // This field is the IP and port on which the request from the user was
+  // received, stored in ipv4:port or [ipv6]:port format.
+  string destination_host = 7;
+
+  // Size of the HTTP request body in bytes
+  google.protobuf.UInt64Value request_body_size = 8;
+
+  // Size of the HTTP response body in bytes
+  google.protobuf.UInt64Value response_body_size = 9;
+
+  // Size of the HTTP request headers in bytes
+  google.protobuf.UInt64Value request_hdrs_size = 10;
```

Comment:
Fun fact: you can't actually call this request_headers_size, because of some collision with proto generated code around the request_headers map (protoc doesn't give an error, but generates noncompilable code).  If it's called request_headers_bytes, no problem!  So making this change also allows the slightly less awkward hdrs -> headers change - thanks.